app/javascript/src/PaymentGateways/braintree/BraintreeForm.jsx

```diff
@@ -0,0 +1,149 @@
+  const [braintreeNonceValue, setBraintreeNonceValue] = useState('')
+  const [billingAddressData, setBillingAddressData] = useState<BillingAddressData>(billingAddress)
+  const [isCardValid, setIsCardValid] = useState(false)
+  const [isFormValid, setIsFormValid] = useState(false)
+  const [formErrors, setFormErrors] = useState(validate(formRef, validationConstraints))
+  const [cardError, setCardError] = useState(null)
+  const [isAwaiting, setIsAwaiting] = useState(false)
+
+  useEffect(() => {
+    const getHostedFieldsInstance = async () => {
+      const HFInstance = await createHostedFieldsInstance(hostedFields, braintreeClient, hostedFieldOptions, setIsCardValid, setCardError)
+      setHostedFieldsInstance(HFInstance)
+    }
+    getHostedFieldsInstance()
+  }, [])
+
+  useEffect(() => {
+    if (braintreeNonceValue && formRef.current) {
+      formRef.current.submit()
+    }
+  }, [braintreeNonceValue])
+
+  useEffect(() => {
+    const formValid = isCardValid && !formErrors && !isAwaiting
+    setIsFormValid(formValid)
+  }, [formErrors, isCardValid, isAwaiting])
+
+  const get3DSecureNonce = async (payload) => {
+    const threeDSecureInstance = await create3DSecureInstance(threeDSecure, braintreeClient)
+    const response = await veryfyCard(threeDSecureInstance, payload, billingAddressData)
+    const error = response.name === 'BraintreeError'
+      ? response.code === 'THREEDS_LOOKUP_VALIDATION_ERROR' ? response.details.originalError.details.originalError.error.message : response.message
+      : null
+    const nonce = response.nonce || null
+
+    return {
+      error,
+      nonce
+    }
+  }
+
+  const clearHostedFields = () => {
+    hostedFieldsInstance.clear('number')
+    hostedFieldsInstance.clear('cvv')
+    hostedFieldsInstance.clear('expirationDate')
+  }
+
+  const validateForm = (event: SyntheticEvent<HTMLFormElement>) => {
+    const validationErrors = validate(event.currentTarget, validationConstraints)
+    setFormErrors(validationErrors)
+  }
+
+  const handleCardError = (error: string) => {
+    setCardError(`Credit card errors found: ${error}. Please correct your CC data.`)
+    clearHostedFields()
+    setIsAwaiting(false)
+  }
+
+  const onSubmit = async (event: SyntheticEvent<HTMLInputElement>) => {
+    event.preventDefault()
+    if (hostedFieldsInstance) {
+      setIsAwaiting(true)
+      const payload = await hostedFieldsInstance.tokenize()
+        .then(payload => payload)
+        .catch(error => console.error(error))
+
+      const response3Dsecure = threeDSecureEnabled ? await get3DSecureNonce(payload) : null
+      if (response3Dsecure && response3Dsecure.error) {
+        handleCardError(response3Dsecure.error)
+        return
+      }
+      const nonce = response3Dsecure ? response3Dsecure.nonce : payload.nonce
+      setBraintreeNonceValue(nonce)
+      setIsAwaiting(false)
+    }
+  }
+
+  return (
+    <form
+      id="customer_form"
+      className="form-horizontal customer"
```

Comment:
We should start using more strict class names. For instance this could be a `braintree-form__form-horizontal` (name-of-component__component-variant). I think this is how the guys at PF do it... or something similar at least. This will basically make our styles more maintainable and less leaky.

Reply:
I agree, we should improve naming, but in this PR I'm keeping all class, id, and attribute names, cause they are used in rails controllers, and I don't want to introduce more changes than the strictly needed.